metacat-connector-hive/src/main/java/com/netflix/metacat/connector/hive/HiveConnectorPartitionService.java

```diff
@@ -254,120 +259,165 @@ public int getPartitionCount(
-                    final String hivePartitionUri = getPartitionUri(hivePartition);
-                    if (partitionUri == null || !partitionUri.equals(hivePartitionUri)) {
-                        existingPartitionIds.add(partitionName);
-                        //the partition exists, we should not do anything for the partition exists
-                        //unless we alterifExists
-                        final Partition existingPartition =
-                            hiveMetacatConverters.fromPartitionInfo(tableInfo, partitionInfo);
-                        if (partitionsSaveRequest.getAlterIfExists()) {
-                            existingPartition.setParameters(hivePartition.getParameters());
-                            existingPartition.setCreateTime(hivePartition.getCreateTime());
-                            existingPartition.setLastAccessTime(hivePartition.getLastAccessTime());
-                            existingHivePartitions.add(existingPartition);
+        //
+        // If either checkIfExists or alterIfExists is true, check to see if any of the partitions already exists.
+        // If it exists and if alterIfExists=false, we will drop it before adding.
+        // If it exists and if alterIfExists=true, we will alter it.
+        //
+        if (partitionsSaveRequest.getCheckIfExists() || partitionsSaveRequest.getAlterIfExists()) {
+            existingPartitionMap = getPartitionsByNames(table, partitionNames);
+        }
+
+        for (PartitionInfo partitionInfo : partitionInfos) {
+            final String partitionName = partitionInfo.getName().getPartitionName();
+            final PartitionHolder existingPartitionHolder = existingPartitionMap.get(partitionName);
+            if (existingPartitionHolder == null) {
+                addedPartitionNames.add(partitionName);
+                addedPartitionInfos.add(partitionInfo);
+            } else {
+                final String partitionUri =
+                    partitionInfo.getSerde() != null ? partitionInfo.getSerde().getUri() : null;
+                final String existingPartitionUri = getPartitionUri(existingPartitionHolder);
+                if (partitionUri == null || !partitionUri.equals(existingPartitionUri)) {
+                    existingPartitionNames.add(partitionName);
+                    //the partition exists, we should not do anything for the partition exists
+                    //unless we alterifExists
+                    if (partitionsSaveRequest.getAlterIfExists()) {
+                        if (partitionInfo.getSerde() == null) {
+                            partitionInfo.setSerde(new StorageInfo());
+                        }
+                        if (partitionInfo.getAudit() == null) {
+                            partitionInfo.setAudit(new AuditInfo());
+                        }
+                        if (existingPartitionHolder.getPartition() != null) {
```

Comment:
Separating it out will make this more complicated since it depends on other variables. Null check has been retained from previous code. I will have to revisit the code before I can remove it since it calls into different implementations of getting the partitions.